walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,70 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Finds time ranges of events, that at least one attendee 
+    // of the requested meeting attends those events.
+    ArrayList<TimeRange> conflictTimeRanges = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      if (conflict(event.getAttendees(), request.getAttendees())) {
+        conflictTimeRanges.add(event.getWhen());
+      }
+    }
+
+    if (conflictTimeRanges.size() == 0) {
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+      }
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Sorts ranges by their start time in ascending order.
+    Collections.sort(conflictTimeRanges, TimeRange.ORDER_BY_START);
+
+    // Merges events in conflictTimeRanges.
+    ArrayList<TimeRange> mergedTimeRanges = new ArrayList<TimeRange>();
+    TimeRange current = conflictTimeRanges.get(0);
+    for (TimeRange range : conflictTimeRanges) {
+      if (range.overlaps(current)) {
+        current = TimeRange.fromStartEnd(current.start(), Math.max(current.end(), range.end()), false);
+      }else {
+        mergedTimeRanges.add(current);
+        current = range;
+      }
+    }
+    mergedTimeRanges.add(current);
+
+    // Finds time ranges that can hold the requested meeting.
+    ArrayList<TimeRange> meetingTimeRanges = new ArrayList<TimeRange>();
+    int start = TimeRange.START_OF_DAY;
+    for (TimeRange range : mergedTimeRanges) {
+      TimeRange freeTimeRange = TimeRange.fromStartEnd(start, range.start(), false);
+      if (freeTimeRange.duration() >= request.getDuration()) {
+        meetingTimeRanges.add(freeTimeRange);
+      }
+      start = range.end();
+    }
+    // Checks the last free time range of the day.
+    TimeRange lastTimeRange = TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true);
+    if (lastTimeRange.duration() >= request.getDuration()) {
+      meetingTimeRanges.add(lastTimeRange);
+    }
+
+    return meetingTimeRanges;
+  }
+
+  // Checks if there is a same attendee in two events.
+  private boolean conflict(Collection<String> attendeesA, Collection<String> attendeesB) {
+    for (String a : attendeesA) {
+      for (String b : attendeesB) {
```

Comment:
you can use attandeesB.contains(a) to check, which is simpler and more efficient

Reply:
I changed my code to use attandeesB.contains(a) to check for conflict.